### PR TITLE
Add code and yaml for Istio as an addon

### PR DESCRIPTION
**Pull request by ostromart**
This commit adds the following:
 - YAML for istio components to the addons directory (versions with
   and without auth)
 - logic to copy the appropriate YAML file into the addons manager
   dir, depending on ENABLE_ISTIO and ISTIO_AUTH_TYPE env vars
 - changes to addon manager script which allow plugins to exist in
   namespace other than kube-system. This is needed for Istio
   functionality
 - additional initialization for Istio in addon manager.

This has been manually tested - e2e tests are in progress but require additional code to enable istio in tests. See:
 - https://github.com/istio/istio/pull/2841
 - https://github.com/kubernetes/test-infra/pull/6612

**What this PR does / why we need it**:
Allow [Istio](https://istio.io/) to be installed as an addon. This simplifies the Istio setup process on k8s and GKE. Istio team is trying to make it easier for end users to get to a basic functional installation.  
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh
Comment (MrHohn) [inline cluster/gce/gci/configure-helper.sh]: Humm, I was under the impression that we also need to pipe these envs into build-kube-env:
https://github.com/kubernetes/kubernetes/blob/9b646503113ae9eb33e57ca231edb868247ba1f5/cluster/common.sh#L472

Does it just work?

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: Remove this if no longer needed?

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: Why do we need this extra step? Are Istio addon manifests not under ${ADDON_PATH}?

File: cluster/addons/istio/auth/istio-auth.yaml
Comment (mikedanese) [inline cluster/addons/istio/auth/istio-auth.yaml]: apikeys?

File: cluster/addons/addon-manager/kube-addons.sh
Comment (ostromart) [reply cluster/addons/addon-manager/kube-addons.sh]: They are there but it's a long manifest and not all of the resources are labeled with CLUSTER_SERVICE_LABEL, only the deployments at the moment so the apply fails for some resources unless they are already created.

File: cluster/gce/gci/configure-helper.sh
Comment (ostromart) [reply cluster/gce/gci/configure-helper.sh]: I manually tested by just temporarily setting the variables directly in the script itself. This will work for GKE which has an internal path for setting env vars but it will be good to implement the non-GKE path as well. I'll take a look at that.

File: cluster/addons/istio/auth/istio-auth.yaml
Comment (ostromart) [reply cluster/addons/istio/auth/istio-auth.yaml]: It does seem to be a typo in Istio repo. Still investigating on Istio side but here we have to use the misspelled version since that's what is in Istio repo at the version we are using here.

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [reply cluster/addons/addon-manager/kube-addons.sh]: Thanks for the clarification, it sounds like we need a way to express the creation order. I wonder if there is a more generic way to achieve this.

Note that the resources without any of `CLUSTER_SERVICE_LABEL` or `ADDON_MANAGER_LABEL` will not be deleted when its manifest is removed from the folder (e.g. when disabling istio). Addon-manager's cleanup logic is accomplished via `kubectl apply -f {FOLDER} --prune=true -l {LABEL}`, which relies on label matching.

Also, I vaguely remember that current implementation of `kubectl apply --prune` doesn't handle the multi-namespaces cases very well. Let's say we have two manifests, one is under namespace1, the other under namespace2. `kubectl apply -f --prune` works well for cleaning up resource as long as there are at least one resource in each namespace defined in manifests. Once the last resource in that namespace gets remove from the manifest (e.g. again, disabling istio), this resource will not be cleaned up because `kubectl` will not even look into that namespace if it is neither given via `--namespace` flag nor defined in the manifests.

I start feeling that the solution would be messy given the inflexibility of addon-manager...

File: cluster/addons/istio/auth/istio-auth.yaml
Comment (ostromart) [reply cluster/addons/istio/auth/istio-auth.yaml]: https://github.com/istio/istio/pull/3101

**Comment by ostromart** [comment]
Per conversation with @MrHohn I've labeled all Istio resources with addonmanager.kubernetes.io/mode: Reconcile, and the Deployments as kubernetes.io/cluster-service: "true" so these show up in kubectl cluster-info. I've removed the special handling for unlabeled resources out of kube-addons.sh so the only change there is now removing the namespace filter.

File: cluster/gce/gci/configure-helper.sh
Comment (MrHohn) [reply cluster/gce/gci/configure-helper.sh]: I see, yeah it would be great to have a common path so that we could enable it in CI jobs as well (for GCE cluster).

File: cluster/addons/istio/noauth/istio.yaml
Comment (MrHohn) [inline cluster/addons/istio/noauth/istio.yaml]: As we discussed before, for enable this seems fine. But for disable, addon-manager will not do the cleanup when all resources under this new namespace are removed from manifests. Is it acceptable for the alpha phase?

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: I think this will work, but I need to confirm how `--prune` behave when multiple namespaces appear in manifest. Will give an update soon.

File: cluster/addons/istio/noauth/istio.yaml
Comment (MrHohn) [inline cluster/addons/istio/noauth/istio.yaml]: Oops sorry, I said it wrong, this label should be attached to the `istio-mixer` service instead.
https://github.com/kubernetes/kubernetes/blob/c322f1d06504fa5b81e117ec79dd77fb4e769b4f/pkg/kubectl/cmd/clusterinfo.go#L75-L80

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [reply cluster/addons/addon-manager/kube-addons.sh]: Seems like it should work with the multiple namespaces case, kubectl records the visited namespaces for pruning:
https://github.com/kubernetes/kubernetes/blob/48388fec7eaad4ac8d84fbe20673ffacf41964a1/pkg/kubectl/cmd/apply.go#L391-L402

Also a bad news, at this moment the object types for pruning is hardcoded so don't think it will work well with the CRDs added by istio...
https://github.com/kubernetes/kubernetes/blob/48388fec7eaad4ac8d84fbe20673ffacf41964a1/pkg/kubectl/cmd/apply.go#L418-L440

File: cluster/addons/istio/noauth/istio.yaml
Comment (ostromart) [reply cluster/addons/istio/noauth/istio.yaml]: Done.

File: cluster/gce/gci/configure-helper.sh
Comment (ostromart) [reply cluster/gce/gci/configure-helper.sh]: https://github.com/kubernetes/test-infra/pull/6612 is still in progress from @krzyzacy 
Let me see if I can help with that. Seems like it would be generally useful functionality. 
The public API and CLI are being updated right now (ETA ~2 weeks) so that would be the way to enable it in tests when it becomes available.

File: cluster/addons/istio/noauth/istio.yaml
Comment (MrHohn) [inline cluster/addons/istio/noauth/istio.yaml]: General comment for all deployments that have `replicas: 1`: would any of them want the ability to horizontally scaled up? If so, you may want to remove this field in manifest so addon-manager will not scale it back down.

File: cluster/addons/istio/noauth/istio.yaml
Comment (MrHohn) [inline cluster/addons/istio/noauth/istio.yaml]: Another general comment: it seems like all podTemplates don't have resource quotas set (memory and cpu), so all pods will be in `BestEffort` class. It may increase the risk of overcommitting the nodes and istio components being easily evicted. For important system pod, a general suggestion would be making it `Guaranteed` class and set a high pod priority. Don't think this is a blocker but would be great to have better resource accounting in future.


Ref https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/, https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/

File: cluster/addons/istio/auth/istio-auth.yaml
Comment (MrHohn) [inline cluster/addons/istio/auth/istio-auth.yaml]: nit: double `---` lines?

File: cluster/addons/istio/noauth/istio.yaml
Comment (ostromart) [reply cluster/addons/istio/noauth/istio.yaml]: This is ok for alpha since we won't be seeing loads that the default sizing can't handle (in fact it's supposed to be able to handle much higher loads). But longer term we need a solution since we do want to reconcile all the other Deployment fields.

File: cluster/addons/istio/noauth/istio.yaml
Comment (ostromart) [reply cluster/addons/istio/noauth/istio.yaml]: I had some discussion in the team meeting about this just now. For alpha, it's ok to go with the BestEffort since we don't yet have a full understanding of performance and scale. This is something that will evolve as we move to 1.0. 
But this is a very good point and we will be tracking this issue.

File: cluster/addons/istio/noauth/istio.yaml
Comment (ostromart) [reply cluster/addons/istio/noauth/istio.yaml]: Per offline discussion, I've removed the replicas setting so it will not be reconciled.

File: cluster/addons/istio/auth/istio-auth.yaml
Comment (ostromart) [reply cluster/addons/istio/auth/istio-auth.yaml]: Done.

**Review by MrHohn — APPROVED** [review]
Looks good, don't have other major comments.
/lgtm

**Comment by ostromart** [comment]
@mikedanese, would you please merge this if there are no outstanding concerns?

**Review by mikedanese — COMMENTED** [review]
Should these manifests go in cluster/gce/addons if they are gke specific?

Can you squash fixup commits?

File: cluster/addons/istio/auth/istio-auth.yaml
Comment (ostromart) [reply cluster/addons/istio/auth/istio-auth.yaml]: 1. Squash: done.
2. The manifests themselves are not GKE specific, the part that's GKE specific is installing the istio manifest based on a startup env var which is set by GKE. But a user can also just copy the manifest through some other means and it will work the same. 
3. Sent this question to istio security team, should have answer shortly.

Comment (ostromart) [reply cluster/addons/istio/auth/istio-auth.yaml]: I've added you to the email exchange on 3. Is this a blocker for this PR or could we open an issue and move ahead?

Comment (ostromart) [reply cluster/addons/istio/auth/istio-auth.yaml]: Just a followup on the secrets issue. Per offline discussion, there are currently only two types of secret that Istio needs to read - however, there's no way to filter based on type so there doesn't seem to be a way to narrow the scope. Unless you have some ideas here I'm not going to pursue this further.

**Pull request merged**
